clang/lib/3C/ArrayBoundsInferenceConsumer.cpp

```diff
@@ -835,50 +882,73 @@ void LengthVarInference::VisitStmt(Stmt *St) {
-      VisitArraySubscriptExpr(SubASE);
-      return;
+  //assert (CurBB != nullptr && "Array dereference does not belong "
+  //                            "to any basic block");
+  // If this does not belong to any BB, ignore.
+  if (CurBB == nullptr)
+    return;
+  // First, get the BoundsKey for the base.
+  Expr *BE = ASE->getBase()->IgnoreParenCasts();
+
+  // If this is a multi-level array dereference i.e., a[i][j],
+  // then try-processing the base ASE i.e., a[i].
+  if (ArraySubscriptExpr *SubASE = dyn_cast_or_null<ArraySubscriptExpr>(BE)) {
+    VisitArraySubscriptExpr(SubASE);
+    return;
+  }
+  //auto BaseCVars = CR->getExprConstraintVars(BE);
+  // Next get the index used.
+  Expr *IdxExpr = ASE->getIdx()->IgnoreParenCasts();
+  //auto IdxCVars = CR->getExprConstraintVars(IdxExpr);
+  BoundsKey BasePtr, IdxKey;
+  auto &ABI = I.getABoundsInfo();
+
+  // Get the bounds key of the base and index.
+  if (tryGetValidBoundsKey(BE, BasePtr, I, C) &&
+      tryGetValidBoundsKey(IdxExpr, IdxKey, I, C)) {
+    std::set<BoundsKey> PossibleLens;
+    PossibleLens.clear();
+    ComparisionVisitor CV(I, C, IdxKey, PossibleLens);
+    auto &CDNodes = CDG->getControlDependencies(CurBB);
+    if (!CDNodes.empty()) {
+      // Next try to find all the nodes that the CurBB is
+      // control dependent on.
+      // For each of the control dependent node, check if we are comparing the
+      // index variable with another variable.
+      for (auto &CDGNode : CDNodes) {
+        // Collect the possible length bounds keys.
+        CV.TraverseStmt(CDGNode->getTerminatorStmt());
+      }
+      ABI.updatePotentialCountBounds(BasePtr, PossibleLens);
     }
-    auto BaseCVars = CR->getExprConstraintVars(BE);
-    // Next get the index used.
-    Expr *IdxExpr = ASE->getIdx()->IgnoreParenCasts();
-    auto IdxCVars = CR->getExprConstraintVars(IdxExpr);
-
-    // Get the bounds key of the base and index.
-    if (CR->containsValidCons(BaseCVars) && !CR->containsValidCons(IdxCVars)) {
-      BoundsKey BasePtr, IdxKey;
-      auto &ABI = I.getABoundsInfo();
-      if (CR->resolveBoundsKey(BaseCVars, BasePtr) &&
-          (CR->resolveBoundsKey(IdxCVars, IdxKey) ||
-           ABI.tryGetVariable(IdxExpr, *C, IdxKey))) {
-        std::set<BoundsKey> PossibleLens;
-        PossibleLens.clear();
-        ComparisionVisitor CV(I, C, IdxKey, PossibleLens);
-        auto &CDNodes = CDG->getControlDependencies(CurBB);
-        if (!CDNodes.empty()) {
-          // Next try to find all the nodes that the CurBB is
-          // control dependent on.
-          // For each of the control dependent node, check if we are comparing
-          // the index variable with another variable.
-          for (auto &CDGNode : CDNodes) {
-            // Collect the possible length bounds keys.
-            CV.TraverseStmt(CDGNode->getTerminatorStmt());
-          }
-          ABI.updatePotentialCountBounds(BasePtr, PossibleLens);
+  }
+
+  /*
```

Comment:
Commented code can probably be removed. It looks more or less the same as the uncommented code above.